File: file2png.go
package main

import (
    "encoding/binary"
    "flag"
    "fmt"
    "image"
    "image/color"
    "image/png"
    "io"
    "math"
    "os"
)

func main() {
    flag.Usage = func() {
        fmt.Fprintf(os.Stderr, "file2png - Convert any file to PNG and back\n\n")
        fmt.Fprintf(os.Stderr, "Usage:\n")
        fmt.Fprintf(os.Stderr, "  Encode: file2png input_file output.png\n")
        fmt.Fprintf(os.Stderr, "  Decode: file2png -d input.png output_file\n")
        fmt.Fprintf(os.Stderr, "  Pipe:   command1 | file2png > output.png\n")
        fmt.Fprintf(os.Stderr, "          file2png -d < input.png > output_file\n\n")
        fmt.Fprintf(os.Stderr, "Options:\n")
        fmt.Fprintf(os.Stderr, "  -d    decode PNG to file\n")
        fmt.Fprintf(os.Stderr, "  -h    show this help message\n\n")
        fmt.Fprintf(os.Stderr, "Examples:\n")
        fmt.Fprintf(os.Stderr, "  file2png document.pdf document.png\n")
        fmt.Fprintf(os.Stderr, "  file2png -d document.png document.pdf\n")
        fmt.Fprintf(os.Stderr, "  cat file | file2png > encoded.png\n")
    }

    var decode bool
    flag.BoolVar(&decode, "d", false, "decode PNG to file")
    flag.Parse()

    args := flag.Args()

    var input io.Reader = os.Stdin
    var output io.Writer = os.Stdout

    // Show usage only if arguments are provided but incorrect
    if len(args) != 0 && len(args) != 2 {
        flag.Usage()
        os.Exit(1)
    }

    if len(args) == 2 {
        inFile, err := os.Open(args[0])
        if err != nil {
            fmt.Fprintf(os.Stderr, "Error opening input file: %v\n", err)
            os.Exit(1)
        }
        defer inFile.Close()
        input = inFile

        outFile, err := os.Create(args[1])
        if err != nil {
            fmt.Fprintf(os.Stderr, "Error creating output file: %v\n", err)
            os.Exit(1)
        }
        defer outFile.Close()
        output = outFile
    }

    if decode {
        if err := decodeFile(input, output); err != nil {
            fmt.Fprintf(os.Stderr, "Error decoding: %v\n", err)
            os.Exit(1)
        }
        return
    }

    if err := encodeFile(input, output); err != nil {
        fmt.Fprintf(os.Stderr, "Error encoding: %v\n", err)
        os.Exit(1)
    }
}

func encodeFile(r io.Reader, w io.Writer) error {
    data, err := io.ReadAll(r)
    if err != nil {
        return err
    }

    fileSize := len(data)
    pixelCount := (fileSize + 8) / 3
    dimension := int(math.Ceil(math.Sqrt(float64(pixelCount))))
    
    img := image.NewRGBA(image.Rect(0, 0, dimension, dimension))

    sizeBuf := make([]byte, 8)
    binary.LittleEndian.PutUint64(sizeBuf, uint64(fileSize))

    x, y := 0, 0

    for i := 0; i < 8; i += 3 {
        r := sizeBuf[i]
        g := byte(0)
        b := byte(0)
        if i+1 < 8 {
            g = sizeBuf[i+1]
        }
        if i+2 < 8 {
            b = sizeBuf[i+2]
        }
        img.Set(x, y, color.RGBA{R: r, G: g, B: b, A: 255})
        x++
        if x >= dimension {
            x = 0
            y++
        }
    }

    for i := 0; i < len(data); i += 3 {
        r := data[i]
        g := byte(0)
        b := byte(0)
        if i+1 < len(data) {
            g = data[i+1]
        }
        if i+2 < len(data) {
            b = data[i+2]
        }
        img.Set(x, y, color.RGBA{R: r, G: g, B: b, A: 255})
        x++
        if x >= dimension {
            x = 0
            y++
        }
    }

    return png.Encode(w, img)
}

func decodeFile(r io.Reader, w io.Writer) error {
    img, err := png.Decode(r)
    if err != nil {
        return err
    }

    bounds := img.Bounds()
    width, height := bounds.Max.X, bounds.Max.Y

    sizeBuf := make([]byte, 8)
    pixelIndex := 0
    for i := 0; i < 8; i += 3 {
        x := pixelIndex % width
        y := pixelIndex / width
        c := img.At(x, y)
        r, g, b, _ := c.RGBA()
        sizeBuf[i] = byte(r)
        if i+1 < 8 {
            sizeBuf[i+1] = byte(g)
        }
        if i+2 < 8 {
            sizeBuf[i+2] = byte(b)
        }
        pixelIndex++
    }

    fileSize := binary.LittleEndian.Uint64(sizeBuf)
    data := make([]byte, fileSize)
    dataIndex := 0

    startPixel := (8 + 2) / 3

    for y := startPixel / width; y < height; y++ {
        startX := 0
        if y == startPixel/width {
            startX = startPixel % width
        }
        for x := startX; x < width; x++ {
            if dataIndex >= int(fileSize) {
                break
            }
            c := img.At(x, y)
            r, g, b, _ := c.RGBA()
            data[dataIndex] = byte(r)
            if dataIndex+1 < int(fileSize) {
                data[dataIndex+1] = byte(g)
            }
            if dataIndex+2 < int(fileSize) {
                data[dataIndex+2] = byte(b)
            }
            dataIndex += 3
        }
    }

    _, err = w.Write(data[:fileSize])
    return err
}

File: file2png.py
import argparse
import math
import os
import struct
from PIL import Image
import sys

def encode_file(input_stream, output_stream):
    data = input_stream.read()
    file_size = len(data)
    pixel_count = (file_size + 8) // 3
    dimension = math.ceil(math.sqrt(pixel_count))

    img = Image.new('RGBA', (dimension, dimension), (0, 0, 0, 255))
    pixels = img.load()

    size_buf = struct.pack('<Q', file_size)
    x, y = 0, 0

    for i in range(0, 8, 3):
        r = size_buf[i]
        g = size_buf[i+1] if i+1 < 8 else 0
        b = size_buf[i+2] if i+2 < 8 else 0
        pixels[x, y] = (r, g, b, 255)
        x += 1
        if x >= dimension:
            x = 0
            y += 1

    for i in range(0, len(data), 3):
        r = data[i]
        g = data[i+1] if i+1 < len(data) else 0
        b = data[i+2] if i+2 < len(data) else 0
        pixels[x, y] = (r, g, b, 255)
        x += 1
        if x >= dimension:
            x = 0
            y += 1

    img.save(output_stream, format='PNG')

def decode_file(input_stream, output_stream):
    img = Image.open(input_stream)
    pixels = img.load()
    width, height = img.size

    size_buf = bytearray(8)
    pixel_index = 0
    for i in range(0, 8, 3):
        x = pixel_index % width
        y = pixel_index // width
        r, g, b, _ = pixels[x, y]
        size_buf[i] = r
        if i+1 < 8:
            size_buf[i+1] = g
        if i+2 < 8:
            size_buf[i+2] = b
        pixel_index += 1

    file_size = struct.unpack('<Q', size_buf)[0]
    data = bytearray(file_size)

    data_index = 0
    start_pixel = (8 + 2) // 3
    for y in range(start_pixel // width, height):
        start_x = 0 if y != start_pixel // width else start_pixel % width
        for x in range(start_x, width):
            if data_index >= file_size:
                break
            r, g, b, _ = pixels[x, y]
            data[data_index] = r
            if data_index + 1 < file_size:
                data[data_index+1] = g
            if data_index + 2 < file_size:
                data[data_index+2] = b
            data_index += 3

    output_stream.write(data[:file_size])

def main():
    parser = argparse.ArgumentParser(description="file2png - Convert any file to PNG and back")
    parser.add_argument("-d", action="store_true", help="decode PNG to file")
    parser.add_argument("input", nargs="?", default="-", help="Input file (default: stdin)")
    parser.add_argument("output", nargs="?", default="-", help="Output file (default: stdout)")

    args = parser.parse_args()

    input_stream = sys.stdin.buffer if args.input == "-" else open(args.input, "rb")
    output_stream = sys.stdout.buffer if args.output == "-" else open(args.output, "wb")

    try:
        if args.d:
            decode_file(input_stream, output_stream)
        else:
            encode_file(input_stream, output_stream)
    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)
    finally:
        if input_stream is not sys.stdin.buffer:
            input_stream.close()
        if output_stream is not sys.stdout.buffer:
            output_stream.close()

if __name__ == "__main__":
    main()
